### PR TITLE
Python api support

### PR DESCRIPTION
**Pull request by megaSpoon**
## What changes were proposed in this pull request?
DLModel, DLClassifer, DLModel, DLClassifier Python wrapper codes
PythonBigDL class (add responsive methods from python wrapper)
Examples to test

## How was this patch tested?

test with writtern examples comparing the results with the scala version

### CODE REVIEW

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala [hunk 1]
```diff
@@ -271,7 +271,7 @@ object DLModel {
 trait HasBatchSize extends Params {
 
   final val batchSize: Param[Int] = new Param[Int](this, "batchSize", "batchSize")
-  setDefault(batchSize -> 1)
+  setDefault(batchSize -> 4)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: different default value in python and scala?

File: pyspark/bigdl/ml_piepeline/dl_classifier.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+from pyspark.ml.param.shared import *
+from bigdl.util.common import *
+from pyspark.ml.wrapper import JavaParams
+
+if sys.version >= '3':
+    long = int
+    unicode = str
+
+
+class HasBatchSize(Params):
```
Comment (jason-dai) [inline pyspark/bigdl/ml_piepeline/dl_classifier.py]: Do we need a HasMaxEpoch class too?

File: pyspark/bigdl/ml_piepeline/dl_classifier.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        return self
+
+    def getBatchSize(self):
+        """
+        Gets the value of batchSize or its default value.
+        """
+        return self.getOrDefault(self.batchSize)
+
+
+class DLEstimator(JavaParams, Estimator, HasFeaturesCol, HasLabelCol, HasPredictionCol, HasBatchSize):
```
Comment (jason-dai) [inline pyspark/bigdl/ml_piepeline/dl_classifier.py]: Can we make it a subclass of JavaValue too? If so, there is no need to duplicate the jvm_class_constructor code below.

File: pyspark/bigdl/ml_piepeline/dl_classifier.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        return name
+
+    def __init__(self,  model, featureSize, jvalue=None, bigdl_type="float"):
+        super(DLModel, self).__init__()
+        self.value = jvalue if jvalue else callBigDlFunc(
+            bigdl_type, self.jvm_class_constructor(), model, featureSize)
+        self.bigdl_type = bigdl_type
+
+    def setFeatureSize(self, value):
+        callBigDlFunc(self.bigdl_type, "dlmodelSetFeatureSize", self.value, value)
```
Comment (jason-dai) [inline pyspark/bigdl/ml_piepeline/dl_classifier.py]: Is dlmodelSetFeatureSize implemented? Maybe we don't need this method.

File: pyspark/bigdl/ml_piepeline/dl_classifier.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        return self
+
+    def getFeatureSize(self):
+        return callBigDlFunc(self.bigdl_type, "dlmodelGetFeatureSize", self.value)
```
Comment (jason-dai) [inline pyspark/bigdl/ml_piepeline/dl_classifier.py]: Is dlmodelGetFeatureSize implemented? Maybe we don't need this method.

**Comment by jason-dai** [comment]
Please add some unit tests (see https://github.com/intel-analytics/BigDL/tree/master/pyspark/test)

**Comment by jason-dai** [comment]
@jenniew @hhbyyh please take a look

File: pyspark/bigdl/ml_piepeline/dl_classifier.py (hunk 3 above)
Comment (megaSpoon) [reply pyspark/bigdl/ml_piepeline/dl_classifier.py]: I delete the` jvm_class_constructor` method declaration, import `JavaValue` from `bigdl.util.common` and replace it with JavaValue.jvm_class_constructor now. JavaValue's init method needs to specify parameters DLEstimator doesn't have like `jvalue` which could be easier by just using the class method outside the JavaValue class.

**Comment by megaSpoon** [comment]
For unit tests, I will use the codes in the LR examples for estimator and classifier to see if the prediction result and expected result are the same. But I have a question about the unit test for estimator since the  prediction result doesn't remain the same each time I run, probably due to the thing that in the training stage, the optimizer initializes some parameters randomly. Thus, what could be the best strategy for writing a unit test but the expected result of the test is unknown.

File: pyspark/bigdl/ml_piepeline/dl_classifier.py (hunk 3 above)
Comment (jason-dai) [reply pyspark/bigdl/ml_piepeline/dl_classifier.py]: that works for now

**Comment by jason-dai** [comment]
In general for Python API we just need to test it has the same result as the corresponding Scala API.

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala (hunk 1 above)
Comment (megaSpoon) [reply spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: changed back to the same

File: pyspark/bigdl/ml_piepeline/dl_classifier.py (hunk 2 above)
Comment (megaSpoon) [reply pyspark/bigdl/ml_piepeline/dl_classifier.py]: added and also create a HasFeatureSize class for DLModel

**Comment by megaSpoon** [comment]
My confusion above related to the DLEstimator API is that when using DLEstimator, the prediction result is even unrepeatable. It cannot be tested by using scala DLEstimator to produce a result A, using python DLEstimator to produce a result B to see if A equals B due to the randomness. I am guessing we can only do it by initializing the same random seed in their optimizers, am I right? I am trying to figure out how to do it.

**Comment by hhbyyh** [comment]
I thought @jenniew already had a PR on this...

Anyway, for unit tests, just some simple tests should be sufficient. maybe something like scala unit tests.

Also we should support Python DLEstimator/DLModel etc. in spark ml pipeline. please add unit test for that.

**Comment by jason-dai** [comment]
@zhichao-li please take a look

File: pyspark/bigdl/ml_piepeline/dl_classifier.py (hunk 3 above)
Comment (zhichao-li) [reply pyspark/bigdl/ml_piepeline/dl_classifier.py]: You can put `None` for `jvalue` if its empty.
The code is duplicate if without inheritance:
```
+    def jvm_class_constructor(self):
+        name = "create" + self.__class__.__name__
+        print("creating: " + name)
+        return name
+
+    def __init__(self,  model, criterion, featureSize, labelSize, bigdl_type="float"):
+        super(DLEstimator, self).__init__()
+        self.value = callBigDlFunc(
+            bigdl_type, self.jvm_class_constructor(), model, criterion, featureSize, labelSize)
+        self.bigdl_type = bigdl_type
```

```

**Comment by megaSpoon** [comment]
Update:
Thanks for @hhbyyh @yiheng's reviews. 
I am using the [LR pipeline example](https://spark.apache.org/docs/2.1.1/ml-pipeline.html#pipeline) in our BigDL fashion to demonstrate our API's feasibility.

**Comment by hhbyyh** [comment]
That's nice to have. For unit test, just compose any Transformers in Spark (e.g. StandardScaler) and DLClassifier in a pipeline (fit and transform) should be sufficient.

File: pyspark/bigdl/ml_piepeline/dl_classifier.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    def getBatchSize(self):
+        """
+        Gets the value of batchSize or its default value.
+        """
+        return self.getOrDefault(self.batchSize)
+
+
+class DLEstimator(JavaParams, Estimator, HasFeaturesCol, HasLabelCol, HasPredictionCol, HasBatchSize):
+    maxEpoch = Param(Params._dummy(), "maxEpoch", "number of max Epoch")
+    def jvm_class_constructor(self):
```
Comment (zhichao-li) [inline pyspark/bigdl/ml_piepeline/dl_classifier.py]: remove jvm_class_constructor?

File: pyspark/bigdl/ml_piepeline/dl_classifier.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    def _fit(self, dataset):
+        #self._transfer_params_to_java()
+        jmodel = callBigDlFunc(self.bigdl_type, "fitEstimator", self.value, dataset)
+        model = DLModel.of(jmodel, self.bigdl_type)
+        return model
+
+
+class DLModel(JavaParams, Model, HasFeaturesCol, HasPredictionCol, HasBatchSize):
+    def jvm_class_constructor(self):
```
Comment (zhichao-li) [inline pyspark/bigdl/ml_piepeline/dl_classifier.py]: remove?

**Comment by zhichao-li** [comment]
@megaSpoon  why there are two dl_classifier.py in this pr?

**Comment by zhichao-li** [comment]
We need to add some unit-tests for this pipline api to ensure they are callable at the first stage.
i.e "test_pipline.py"

File: pyspark/bigdl/ml_piepeline/dl_classifier.py (hunk 3 above)
Comment (megaSpoon) [reply pyspark/bigdl/ml_piepeline/dl_classifier.py]: already fixed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 8]
```diff
@@ -1976,4 +1978,76 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
...
+
+  def fitClassifier(classifier: DLClassifier[T], dataSet: DataFrame): DLModel[T] = {
+    classifier.fit(dataSet)
+  }
+
+  def setBatchSizeDLEstimator(estimator: DLEstimator[T], batchSize: Int): DLEstimator[T] = {
+    estimator.setBatchSize(batchSize)
+  }
+
+  def setBatchSizeDLClassifier(classifier: DLClassifier[T], batchSize: Int): DLClassifier[T] = {
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: For such simple setter, actually we can invoke it directly from python without adding java wrapper method.
but anyway, adding a java wrapper is ok as well.
i.e
```
    def set_name(self, name):
        """
        Give this model a name. There would be a generated name
        consist of class name and UUID if user doesn't set it.
        """
        callJavaFunc(get_spark_context(), self.value.setName, name)
        return self
```

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 9]
```diff
@@ -1979,6 +1979,12 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
     module.quantize()
   }
 
+  def createDLEstimator() = {
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: empty method?

File: .gitignore [hunk 10]
```diff
@@ -35,4 +35,8 @@ nohup.out
 .ipynb_checkpoints/
 pyspark/dist/
 pyspark/build/
-
+pyspark/bigdl/models/ml_pipeline/*.ipynb_checkpoints/
+pyspark/bigdl/models/ml_pipeline/metastore_db
+pyspark/bigdl/models/ml_pipeline/derby.log
+pyspark/bigdl/models/ml_pipeline/*.ipynb
+pyspark/bigdl/models/ml_pipeline/run.sh
```
Comment (zhichao-li) [inline .gitignore]: useless for the other user?

File: pyspark/test/bigdl/test_dl_classifier.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+from bigdl.util.common import *
+from bigdl.models.ml_pipeline.dl_classifier import *
+from pyspark.sql.types import *
+from pyspark.context import SparkContext
+import numpy as np
+import pytest
+from numpy.testing import assert_allclose
+
+
+class TestTensorflow():
```
Comment (jason-dai) [inline pyspark/test/bigdl/test_dl_classifier.py]: why TestTensorflow?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 12]
```diff
@@ -1979,6 +1979,12 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
...
+  }
+
+  def createDLClassifier() = {
+  }
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: Are the above two methods used?

File: pyspark/test/bigdl/test_dl_classifier.py (hunk 11 above)
Comment (megaSpoon) [reply pyspark/test/bigdl/test_dl_classifier.py]: naming error, fixed!

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala (hunk 12 above)
Comment (megaSpoon) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: removed, originally used to test the javaValue.jvm_class_constructor

File: pyspark/test/bigdl/test_dl_classifier.py (hunk 11 above)
Comment (megaSpoon) [reply pyspark/test/bigdl/test_dl_classifier.py]: I run the run-all.sh in the pyspark/test/dev folder. It passed without error.

Comment (megaSpoon) [reply pyspark/test/bigdl/test_dl_classifier.py]: @zhichao-li Is this the way above to run tests in pyspark/test/bigdl folder?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 13]
```diff
@@ -1976,4 +1978,76 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
...
+    classifier.setBatchSize(batchSize)
+  }
+
+  def setMaxEpochDLEstimator(estimator: DLEstimator[T], maxEpoch: Int): DLEstimator[T] = {
+    estimator.setMaxEpoch(maxEpoch)
+  }
+
+  def setMaxEpochDLClassifier(classifier: DLClassifier[T], maxEpoch: Int): DLClassifier[T] = {
+    classifier.setMaxEpoch(maxEpoch)
+  }
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: Is it possible to combine the above two methods?

**Comment by jason-dai** [comment]
Please rebase and run a Jenkins validation test

File: spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala [hunk 14]
```diff
@@ -91,7 +91,7 @@ class DLEstimator[@specialized(Float, Double) T: ClassTag](
 
   /**
    * learning rate for the optimizer.
-   * Default: 1.0
+   * Default: 0.01
```
Comment (jason-dai) [inline spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: the comment is incorrect?

Comment (megaSpoon) [reply spark/dl/src/main/scala/org/apache/spark/ml/DLEstimator.scala]: change to 0.001 fixed.

**Comment by megaSpoon** [comment]
The unit test #1429 on Jenkins already passed. The PR is ready to merge. @jason-dai

**Comment by megaSpoon** [comment]
I will simplify the codes of set and get methods tomorrow. Test if combining the two set methods together as one or simply invoking them directly from python based on @zhichao-li 's advice works.
